File: app/src/main/java/bb/incognito/utils/TouchHelper.java
package bb.incognito.utils;
import android.graphics.Canvas;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

public class TouchHelper extends ItemTouchHelper.Callback {

    public TouchHelper() {
    }

    public static final float ALPHA_FULL = 1.0f;

    @Override
    public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        final int dragFlags = 0;//ItemTouchHelper.UP | ItemTouchHelper.DOWN;
        final int swipeFlags = ItemTouchHelper.START | ItemTouchHelper.END;
        return makeMovementFlags(dragFlags, swipeFlags);
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        //NoteAdapter.getInstance().onItemDismiss(viewHolder.getAdapterPosition());

    }

    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder source, RecyclerView.ViewHolder target) {
        if (source.getItemViewType() != target.getItemViewType()) {
            return false;
        }
        //NoteAdapter.getInstance().onItemMove(source.getAdapterPosition(), target.getAdapterPosition());
        return true;
    }

    @Override
    public void onChildDraw(Canvas c, RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
        if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
            final float alpha = ALPHA_FULL - Math.abs(dX) / (float) viewHolder.itemView.getWidth();
            viewHolder.itemView.setAlpha(alpha);
            viewHolder.itemView.setTranslationX(dX);
        } else {
            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
        }
    }

    @Override
    public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
//            if (viewHolder instanceof NoteViewHolder) {
//                // Let the view holder know that this item is being moved or dragged
//                NoteViewHolder itemViewHolder = (NoteViewHolder) viewHolder;
//                itemViewHolder.onItemSelected();
//            }
        }

        super.onSelectedChanged(viewHolder, actionState);
    }

    @Override
    public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        super.clearView(recyclerView, viewHolder);

        viewHolder.itemView.setAlpha(ALPHA_FULL);

//        if (viewHolder instanceof NoteViewHolder) {
//            // Tell the view holder it's time to restore the idle state
//            NoteViewHolder itemViewHolder = (NoteViewHolder) viewHolder;
//            itemViewHolder.onItemClear();
//        }
    }
}

File: app/src/main/java/bb/incognito/utils/TouchHelper.kt
package bb.incognito.utils

import android.graphics.Canvas
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper

class TouchHelper : ItemTouchHelper.Callback() {

    override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
        val dragFlags = 0//ItemTouchHelper.UP | ItemTouchHelper.DOWN;
        val swipeFlags = ItemTouchHelper.START or ItemTouchHelper.END
        return ItemTouchHelper.Callback.makeMovementFlags(dragFlags, swipeFlags)
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        //NoteAdapter.getInstance().onItemDismiss(viewHolder.getAdapterPosition());

    }

    override fun onMove(recyclerView: RecyclerView, source: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
        return if (source.itemViewType != target.itemViewType) {
            false
        } else true
        //NoteAdapter.getInstance().onItemMove(source.getAdapterPosition(), target.getAdapterPosition());
    }

    override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
        if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
            val alpha = ALPHA_FULL - Math.abs(dX) / viewHolder.itemView.width.toFloat()
            viewHolder.itemView.alpha = alpha
            viewHolder.itemView.translationX = dX
        } else {
            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
        }
    }

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
            //            if (viewHolder instanceof NoteViewHolder) {
            //                // Let the view holder know that this item is being moved or dragged
            //                NoteViewHolder itemViewHolder = (NoteViewHolder) viewHolder;
            //                itemViewHolder.onItemSelected();
            //            }
        }

        super.onSelectedChanged(viewHolder, actionState)
    }

    override fun clearView(recyclerView: RecyclerView?, viewHolder: RecyclerView.ViewHolder) {
        super.clearView(recyclerView, viewHolder)

        viewHolder.itemView.alpha = ALPHA_FULL

        //        if (viewHolder instanceof NoteViewHolder) {
        //            // Tell the view holder it's time to restore the idle state
        //            NoteViewHolder itemViewHolder = (NoteViewHolder) viewHolder;
        //            itemViewHolder.onItemClear();
        //        }
    }

    companion object {

        val ALPHA_FULL = 1.0f
    }
}